maintenance/githubutil/githubutil.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package githubutil
+
+import (
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+
+	"github.com/google/go-github/github"
+	"golang.org/x/oauth2"
+)
+
+type repositoryService interface {
+	CreateStatus(org, repo, ref string, status *github.RepoStatus) (*github.RepoStatus, *github.Response, error)
+	GetCombinedStatus(org, repo, ref string, opt *github.ListOptions) (*github.CombinedStatus, *github.Response, error)
+}
+
+type pullRequestService interface {
+	List(org, repo string, opt *github.PullRequestListOptions) ([]*github.PullRequest, *github.Response, error)
+}
+
+// Client is an augmentation of the go-github client that adds retry logic, rate limiting, and pagination
+// handling to some of the client functions.
+type Client struct {
+	prService   pullRequestService
+	repoService repositoryService
+
+	retries             int
+	retryInitialBackoff time.Duration
+
+	tokenReserve int
+	dryRun       bool
+}
+
+// NewClient makes a new githubutil client that does rate limiting and retries.
+func NewClient(token string, dryRun bool) *Client {
+	httpClient := &http.Client{
+		Transport: &oauth2.Transport{
+			Base:   http.DefaultTransport,
+			Source: oauth2.ReuseTokenSource(nil, oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})),
+		},
+	}
+	client := github.NewClient(httpClient)
+	return &Client{
+		prService:           client.PullRequests,
+		repoService:         client.Repositories,
+		retries:             5,
+		retryInitialBackoff: time.Duration(3) * time.Second,
+		tokenReserve:        50,
+		dryRun:              dryRun,
+	}
+}
+
+func (c *Client) retryDelay(retryCount int) {
+	time.Sleep(c.retryInitialBackoff * time.Duration(math.Exp2(float64(retryCount))))
+}
+
+func (c *Client) limitRate(r *github.Rate) {
+	if r.Remaining <= c.tokenReserve {
+		sleepDuration := r.Reset.Sub(time.Now()) + (time.Second * time.Duration(10))
+		if sleepDuration > time.Duration(0) {
+			glog.Infof("--Rate Limiting-- Tokens reached minimum reserve %d. Sleeping until reset in %v.\n", c.tokenReserve, sleepDuration)
+			time.Sleep(sleepDuration)
+		}
+	}
+}
+
+// CreateStatus creates or updates a status context on the indicated reference.
+// This function limits rate and does retries if needed.
```

Comment:
Pull the rate limit and retry logic into its own function so that you're not just copying it three times.